ddl/ddl_api.go

```diff
@@ -5245,48 +5245,66 @@ func checkPlacementSpecConstraint(rules []*placement.Rule, rule *placement.Rule,
 			return rules, err
 		}
 
+		rulesLen := len(rules)
+		ruleCnt := rule.Count
 		for labels, cnt := range constraints {
-			newRule := &placement.Rule{}
+			newRule := &placement.RuleOp{}
 			*newRule = *rule
+			newRule.Rule = &placement.Rule{}
+			*newRule.Rule = *rule.Rule
 			if cnt <= 0 {
-				err = errors.Errorf("count should be non-positive, but got %d", cnt)
+				err = errors.Errorf("count should be positive, but got %d", cnt)
+				break
+			}
+			ruleCnt -= cnt
+			if ruleCnt < 0 {
+				rules = rules[:rulesLen]
+				err = errors.Errorf("COUNT should be larger or equal to the number of total replicas, but got %d", rule.Count)
 				break
 			}
-			// TODO: handle or remove rule.Count in later commits
-			rule.Count -= cnt
 			newRule.Count = cnt
 			err = checkPlacementLabelConstraints(newRule, strings.Split(strings.TrimSpace(labels), ","))
 			if err != nil {
 				break
 			}
 			rules = append(rules, newRule)
 		}
+		rule.Count = ruleCnt
+
+		if rule.Count > 0 {
+			// TODO: currently, PD has no API to update rules
+			// adding rules by a dict with a implicit default rule should increment the replicas number of the default rule.
+			// so TiDB needs to know how many default replicas there is, which requires an HTTP request
+			// for now, it acts like altering rules, i.e., override the default rule
+			// maybe unsupport this feature, or handle 'get rules from PD' smoothly in later PRs
+			rules = append(rules, rule)
+		}
 	} else {
 		err = errors.Errorf("constraint should be a JSON array or object, but got '%s'", cnstr)
 	}
 	return rules, err
 }
 
-func checkPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
-	rules := make([]*placement.Rule, 0, len(specs))
+func checkPlacementSpecs(specs []*ast.PlacementSpec, tableID, partitionID int64) ([]*placement.RuleOp, error) {
```

Comment:
```suggestion
func buildPlacementRuleOp(specs []*ast.PlacementSpec, tableID, partitionID int64) ([]*placement.RuleOp, error) {
```
It's not just checking any more.